File: Diagram/StateClassShape.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2009-2015 ApexRMS.
'
'************************************************************************************

Imports System.Drawing
Imports SyncroSim.Core
Imports SyncroSim.Common.Forms

Class StateClassShape
    Inherits BoxDiagramShape

    Private m_Project As Project
    Private m_StratumIdSource As Nullable(Of Integer)
    Private m_StateClassIdSource As Integer
    Private m_StratumIdDest As Nullable(Of Integer)
    Private m_StateClassIdDest As Nullable(Of Integer)
    Private m_AgeMinimum As Nullable(Of Integer)
    Private m_AgeMaximum As Nullable(Of Integer)
    Private m_SLXId As Integer
    Private m_SLXDisplayName As String
    Private m_SLYId As Integer
    Private m_SLYDisplayName As String
    Private m_TooltipText As String
    Private m_IncomingDT As New List(Of DeterministicTransition)
    Private m_IncomingDTLines As New List(Of DeterministicTransitionLine)
    Private m_OutgoingDTLines As New List(Of DeterministicTransitionLine)
    Private m_IncomingPT As New List(Of Transition)
    Private m_IncomingPTLines As New List(Of ProbabilisticTransitionLine)
    Private m_OutgoingPT As New List(Of Transition)
    Private m_OutgoingPTLines As New List(Of ProbabilisticTransitionLine)
    Private m_StaticBorderPen As New Pen(TRANSITION_DIAGRAM_SHAPE_READONLY_BORDER_COLOR, ZOOM_SAFE_PEN_WIDTH)
    Private m_StaticBkBrush As New SolidBrush(TRANSITION_DIAGRAM_SHAPE_BACKGROUND_COLOR)
    Private m_SharesLocation As Boolean
    Private m_Monitor As DataSheetMonitor
    Private m_IsDisposed As Boolean

    Const ZOOM_SAFE_PEN_WIDTH As Integer = -1

    Public Sub New(
        ByVal project As Project,
        ByVal stratumIdSource As Nullable(Of Integer),
        ByVal stateClassIdSource As Integer,
        ByVal stratumIdDestination As Nullable(Of Integer),
        ByVal stateClassIdDestination As Nullable(Of Integer),
        ByVal ageMinimum As Nullable(Of Integer),
        ByVal ageMaximum As Nullable(Of Integer),
        ByVal stateLabelXId As Integer,
        ByVal stateLabelXName As String,
        ByVal stateLabelXDisplayName As String,
        ByVal stateLabelYId As Integer,
        ByVal stateLabelYDisplayName As String)

        MyBase.New(
            TRANSITION_DIAGRAM_NUM_VERTICAL_CONNECTORS,
            TRANSITION_DIAGRAM_NUM_HORIZONTAL_CONNECTORS)

        Me.m_Project = project
        Me.TitleBarText = stateLabelXName
        Me.m_StratumIdSource = stratumIdSource
        Me.m_StateClassIdSource = stateClassIdSource
        Me.m_StratumIdDest = stratumIdDestination
        Me.m_StateClassIdDest = stateClassIdDestination
        Me.m_AgeMinimum = ageMinimum
        Me.m_AgeMaximum = ageMaximum
        Me.m_SLXId = stateLabelXId
        Me.m_SLXDisplayName = stateLabelXDisplayName
        Me.m_SLYId = stateLabelYId
        Me.m_SLYDisplayName = stateLabelYDisplayName
        Me.TitleHeight = TRANSITION_DIAGRAM_TITLE_BAR_HEIGHT
        Me.ItemHeight = TRANSITION_DIAGRAM_ITEM_HEIGHT
        Me.BackgroundColor = TRANSITION_DIAGRAM_SHAPE_BACKGROUND_COLOR
        Me.SelectedBackgroundColor = Me.BackgroundColor
        Me.TitleBackgroundColor = Me.BackgroundColor
        Me.TitleSelectedBackgroundColor = Me.BackgroundColor
        Me.TitleTextColor = TRANSITION_DIAGRAM_TEXT_COLOR
        Me.TitleSelectedTextColor = TRANSITION_DIAGRAM_SELECTED_TEXT_COLOR
        Me.BorderColor = TRANSITION_DIAGRAM_SHAPE_BORDER_COLOR
        Me.DrawItemSeparators = False

        Me.m_Monitor = New DataSheetMonitor(
            Me.m_Project, DATASHEET_TERMINOLOGY_NAME, AddressOf Me.OnTerminologyChanged)

        Me.m_Monitor.Invoke()

    End Sub

    Protected Overrides Sub Dispose(disposing As Boolean)

        If (disposing And Not Me.m_IsDisposed) Then

            If (Me.m_Monitor IsNot Nothing) Then
                Me.m_Monitor.Dispose()
            End If

            If (Me.m_StaticBorderPen IsNot Nothing) Then
                Me.m_StaticBorderPen.Dispose()
            End If

            If (Me.m_StaticBkBrush IsNot Nothing) Then
                Me.m_StaticBkBrush.Dispose()
            End If

            Me.m_IsDisposed = True

        End If

        MyBase.Dispose(disposing)

    End Sub

    Public ReadOnly Property StratumIdSource() As Nullable(Of Integer)
        Get
            Return Me.m_StratumIdSource
        End Get
    End Property

    Public ReadOnly Property StateClassIdSource() As Integer
        Get
            Return Me.m_StateClassIdSource
        End Get
    End Property

    Public ReadOnly Property StratumIdDest As Nullable(Of Integer)
        Get
            Return Me.m_StratumIdDest
        End Get
    End Property

    Public ReadOnly Property StateClassIdDest() As Nullable(Of Integer)
        Get
            Return Me.m_StateClassIdDest
        End Get
    End Property

    Public ReadOnly Property AgeMinimum() As Nullable(Of Integer)
        Get
            Return Me.m_AgeMinimum
        End Get
    End Property

    Public ReadOnly Property AgeMaximum() As Nullable(Of Integer)
        Get
            Return Me.m_AgeMaximum
        End Get
    End Property

    Public ReadOnly Property StateLabelXId() As Integer
        Get
            Return Me.m_SLXId
        End Get
    End Property

    Public ReadOnly Property StateLabelYId() As Integer
        Get
            Return Me.m_SLYId
        End Get
    End Property

    Public ReadOnly Property OutgoingPT() As List(Of Transition)
        Get
            Return Me.m_OutgoingPT
        End Get
    End Property

    Public ReadOnly Property OutgoingPTLines As List(Of ProbabilisticTransitionLine)
        Get
            Return Me.m_OutgoingPTLines
        End Get
    End Property

    Public ReadOnly Property IncomingPT() As List(Of Transition)
        Get
            Return Me.m_IncomingPT
        End Get
    End Property

    Public ReadOnly Property IncomingPTLines As List(Of ProbabilisticTransitionLine)
        Get
            Return Me.m_IncomingPTLines
        End Get
    End Property

    Public ReadOnly Property IncomingDT() As List(Of DeterministicTransition)
        Get
            Return Me.m_IncomingDT
        End Get
    End Property

    Public ReadOnly Property IncomingDTLines As List(Of DeterministicTransitionLine)
        Get
            Return Me.m_IncomingDTLines
        End Get
    End Property

    Public ReadOnly Property OutgoingDTLines As List(Of DeterministicTransitionLine)
        Get
            Return Me.m_OutgoingDTLines
        End Get
    End Property

    Public WriteOnly Property SharesLocation As Boolean
        Set(value As Boolean)
            Me.m_SharesLocation = value
        End Set
    End Property

    Public Overrides Function GetToolTipText() As String
        Return Me.m_TooltipText
    End Function

    Public Overrides Sub Render(g As Drawing.Graphics)

        If (Me.m_SharesLocation And Not Me.IsSelected) Then

            Dim rc As New Rectangle(
                Me.Bounds.Left - 8,
                Me.Bounds.Top - 8,
                Me.Bounds.Width,
                Me.Bounds.Height)

            g.FillRectangle(Me.m_StaticBkBrush, rc)
            g.DrawRectangle(Me.m_StaticBorderPen, rc)

        End If

        MyBase.Render(g)

    End Sub

    Private Sub OnTerminologyChanged(ByVal e As DataSheetMonitorEventArgs)

        Dim slxlabel As String = Nothing
        Dim slylabel As String = Nothing

        Dim ds As DataSheet = Me.m_Project.GetDataSheet(DATASHEET_TERMINOLOGY_NAME)
        GetStateLabelTerminology(ds, slxlabel, slylabel)

        Me.m_TooltipText =
            slxlabel & ": " & Me.m_SLXDisplayName & vbCrLf &
            slylabel & ": " & Me.m_SLYDisplayName

    End Sub

End Class

File: src/Diagram/StateClassShape.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2021 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System;
using System.Drawing;
using SyncroSim.Core;
using SyncroSim.Common.Forms;
using System.Collections.Generic;

namespace SyncroSim.STSim
{
    internal class StateClassShape : BoxDiagramShape
    {
        private Project m_Project;
        private int? m_StratumIdSource;
        private int m_StateClassIdSource;
        private int? m_StratumIdDest;
        private int? m_StateClassIdDest;
        private int? m_AgeMinimum;
        private int? m_AgeMaximum;
        private int m_SLXId;
        private string m_SLXDisplayName;
        private int m_SLYId;
        private string m_SLYDisplayName;
        private string m_TooltipText;
        private List<DeterministicTransition> m_IncomingDT = new List<DeterministicTransition>();
        private List<DeterministicTransitionLine> m_IncomingDTLines = new List<DeterministicTransitionLine>();
        private List<DeterministicTransitionLine> m_OutgoingDTLines = new List<DeterministicTransitionLine>();
        private List<Transition> m_IncomingPT = new List<Transition>();
        private List<ProbabilisticTransitionLine> m_IncomingPTLines = new List<ProbabilisticTransitionLine>();
        private List<Transition> m_OutgoingPT = new List<Transition>();
        private List<ProbabilisticTransitionLine> m_OutgoingPTLines = new List<ProbabilisticTransitionLine>();
        private Pen m_StaticBorderPen = new Pen(Constants.TRANSITION_DIAGRAM_SHAPE_READONLY_BORDER_COLOR, ZOOM_SAFE_PEN_WIDTH);
        private SolidBrush m_StaticBkBrush = new SolidBrush(Constants.TRANSITION_DIAGRAM_SHAPE_BACKGROUND_COLOR);
        private bool m_SharesLocation;
        private DataSheetMonitor m_Monitor;
        private bool m_IsDisposed;
        private const int ZOOM_SAFE_PEN_WIDTH = -1;

        public StateClassShape(Project project, int? stratumIdSource, int stateClassIdSource, int? stratumIdDestination, int? stateClassIdDestination, int? ageMinimum, int? ageMaximum, int stateLabelXId, string stateLabelXName, string stateLabelXDisplayName, int stateLabelYId, string stateLabelYDisplayName) : base(Constants.TRANSITION_DIAGRAM_NUM_VERTICAL_CONNECTORS, Constants.TRANSITION_DIAGRAM_NUM_HORIZONTAL_CONNECTORS)
        {
            this.m_Project = project;
            this.TitleBarText = stateLabelXName;
            this.m_StratumIdSource = stratumIdSource;
            this.m_StateClassIdSource = stateClassIdSource;
            this.m_StratumIdDest = stratumIdDestination;
            this.m_StateClassIdDest = stateClassIdDestination;
            this.m_AgeMinimum = ageMinimum;
            this.m_AgeMaximum = ageMaximum;
            this.m_SLXId = stateLabelXId;
            this.m_SLXDisplayName = stateLabelXDisplayName;
            this.m_SLYId = stateLabelYId;
            this.m_SLYDisplayName = stateLabelYDisplayName;
            this.TitleHeight = Constants.TRANSITION_DIAGRAM_TITLE_BAR_HEIGHT;
            this.ItemHeight = Constants.TRANSITION_DIAGRAM_ITEM_HEIGHT;
            this.BackgroundColor = Constants.TRANSITION_DIAGRAM_SHAPE_BACKGROUND_COLOR;
            this.SelectedBackgroundColor = this.BackgroundColor;
            this.TitleBackgroundColor = this.BackgroundColor;
            this.TitleSelectedBackgroundColor = this.BackgroundColor;
            this.TitleTextColor = Constants.TRANSITION_DIAGRAM_TEXT_COLOR;
            this.TitleSelectedTextColor = Constants.TRANSITION_DIAGRAM_SELECTED_TEXT_COLOR;
            this.BorderColor = Constants.TRANSITION_DIAGRAM_SHAPE_BORDER_COLOR;
            this.DrawItemSeparators = false;

            this.m_Monitor = new DataSheetMonitor(this.m_Project, Strings.DATASHEET_TERMINOLOGY_NAME, this.OnTerminologyChanged);
            this.m_Monitor.Invoke();
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing && !this.m_IsDisposed)
            {
                if (this.m_Monitor != null)
                {
                    this.m_Monitor.Dispose();
                }

                if (this.m_StaticBorderPen != null)
                {
                    this.m_StaticBorderPen.Dispose();
                }

                if (this.m_StaticBkBrush != null)
                {
                    this.m_StaticBkBrush.Dispose();
                }

                this.m_IsDisposed = true;
            }

            base.Dispose(disposing);
        }

        public int? StratumIdSource
        {
            get
            {
                return this.m_StratumIdSource;
            }
        }

        public int StateClassIdSource
        {
            get
            {
                return this.m_StateClassIdSource;
            }
        }

        public int? StratumIdDest
        {
            get
            {
                return this.m_StratumIdDest;
            }
        }

        public int? StateClassIdDest
        {
            get
            {
                return this.m_StateClassIdDest;
            }
        }

        public int? AgeMinimum
        {
            get
            {
                return this.m_AgeMinimum;
            }
        }

        public int? AgeMaximum
        {
            get
            {
                return this.m_AgeMaximum;
            }
        }

        public int StateLabelXId
        {
            get
            {
                return this.m_SLXId;
            }
        }

        public int StateLabelYId
        {
            get
            {
                return this.m_SLYId;
            }
        }

        public List<Transition> OutgoingPT
        {
            get
            {
                return this.m_OutgoingPT;
            }
        }

        public List<ProbabilisticTransitionLine> OutgoingPTLines
        {
            get
            {
                return this.m_OutgoingPTLines;
            }
        }

        public List<Transition> IncomingPT
        {
            get
            {
                return this.m_IncomingPT;
            }
        }

        public List<ProbabilisticTransitionLine> IncomingPTLines
        {
            get
            {
                return this.m_IncomingPTLines;
            }
        }

        public List<DeterministicTransition> IncomingDT
        {
            get
            {
                return this.m_IncomingDT;
            }
        }

        public List<DeterministicTransitionLine> IncomingDTLines
        {
            get
            {
                return this.m_IncomingDTLines;
            }
        }

        public List<DeterministicTransitionLine> OutgoingDTLines
        {
            get
            {
                return this.m_OutgoingDTLines;
            }
        }

        public bool SharesLocation
        {
            set
            {
                this.m_SharesLocation = value;
            }
        }

        public override string GetToolTipText()
        {
            return this.m_TooltipText;
        }

        public override void Render(System.Drawing.Graphics g)
        {
            if (this.m_SharesLocation && !this.IsSelected)
            {
                Rectangle rc = new Rectangle(this.Bounds.Left - 8, this.Bounds.Top - 8, this.Bounds.Width, this.Bounds.Height);

                g.FillRectangle(this.m_StaticBkBrush, rc);
                g.DrawRectangle(this.m_StaticBorderPen, rc);
            }

            base.Render(g);
        }

        private void OnTerminologyChanged(DataSheetMonitorEventArgs e)
        {
            string slxlabel = null;
            string slylabel = null;

            DataSheet ds = this.m_Project.GetDataSheet(Strings.DATASHEET_TERMINOLOGY_NAME);
            TerminologyUtilities.GetStateLabelTerminology(ds, ref slxlabel, ref slylabel);

            this.m_TooltipText = slxlabel + ": " + this.m_SLXDisplayName + Environment.NewLine + slylabel + ": " + this.m_SLYDisplayName;
        }
    }
}
